diff --git a/CMakeLists.txt b/CMakeLists.txt
index ec782239..57566c5a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -488,7 +488,7 @@
   # https://cmake.org/cmake/help/latest/variable/MSVC_VERSION.html
   (WIN32 AND MSVC_VERSION GREATER 1800))
   if(WIN32)
-    set(GENERATION_SCRIPT py scripts/generate_code.py)
+    set(GENERATION_SCRIPT ${PYTHON_EXECUTABLE} scripts/generate_code.py)
    else()
     set(GENERATION_SCRIPT scripts/generate_code.py)
   endif()
