diff --git a/mlir/CMakeLists.txt b/mlir/CMakeLists.txt
index cd4f05fa05bc..985d5f650b4d 100644
--- a/mlir/CMakeLists.txt
+++ b/mlir/CMakeLists.txt
@@ -77,8 +77,6 @@ include_directories( ${MLIR_INCLUDE_DIR})
 add_subdirectory(include/mlir)
 add_subdirectory(lib)
 add_subdirectory(tools)
-add_subdirectory(unittests)
-add_subdirectory(test)
 
 if( LLVM_INCLUDE_EXAMPLES )
   add_subdirectory(examples)
