diff --git a/CMakeLists.txt b/CMakeLists.txt
index fcd9024..0662e11 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,18 +13,7 @@ else()
   message(STATUS "Build type is '${CMAKE_BUILD_TYPE}'")
 endif()
 
-include(CheckCSourceCompiles)
-check_c_source_compiles(
-  "#if defined(__EMSCRIPTEN__)
-   int main() {return 0;}
-   #endif"
-  BROTLI_EMSCRIPTEN
-)
-if (BROTLI_EMSCRIPTEN)
-  message("-- Compiler is EMSCRIPTEN")
-else()
-  message("-- Compiler is not EMSCRIPTEN")
-endif()
+set(BROTLI_EMSCRIPTEN OFF)
 
 # If Brotli is being bundled in another project, we don't want to
 # install anything.  However, we want to let people override this, so
