diff --git a/cmake/CMakeLists.txt b/cmake/CMakeLists.txt
index 1834422..61cab5b 100644
--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -66,7 +66,7 @@ endif()
 enable_testing()
 
 add_library(port INTERFACE)
-add_library(descriptor_upbproto INTERFACE)
+add_library(descriptor_upb_proto INTERFACE)
 add_library(upb
   ../upb/decode.c
   ../upb/decode_internal.h
@@ -141,15 +141,38 @@ install(
   PATTERN "*.int.h"
 )
 target_include_directories(upb PUBLIC $<INSTALL_INTERFACE:include>)
+
+add_library(all_libs INTERFACE)
+target_link_libraries(all_libs
+  INTERFACE
+  port
+  descriptor_upb_proto
+  upb
+  fastdecode
+  reflection
+  textformat
+  json
+  table
+)
+
+set_target_properties(port PROPERTIES OUTPUT_NAME upb_port)
+set_target_properties(descriptor_upb_proto PROPERTIES OUTPUT_NAME upb_descriptor_upb_proto)
+set_target_properties(fastdecode PROPERTIES OUTPUT_NAME upb_fastdecode)
+set_target_properties(reflection PROPERTIES OUTPUT_NAME upb_reflection)
+set_target_properties(textformat PROPERTIES OUTPUT_NAME upb_textformat)
+set_target_properties(json PROPERTIES OUTPUT_NAME upb_json)
+set_target_properties(table PROPERTIES OUTPUT_NAME upb_table)
+
 install(TARGETS
   port
-  descriptor_upbproto
+  descriptor_upb_proto
   upb
   fastdecode
   reflection
   textformat
   json
   table
+  all_libs
   EXPORT upb-config
 )
 install(EXPORT upb-config NAMESPACE upb:: DESTINATION share/upb)
