diff --git a/cmake/CMakeLists.txt b/cmake/CMakeLists.txt
index d669bf9..1834422 100644
--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -59,6 +59,10 @@ elseif(UNIX)
   set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--build-id")
 endif()
 
+if (MSVC)
+  add_compile_options(/wd4146 /wd4703 -D_CRT_SECURE_NO_WARNINGS)
+endif()
+
 enable_testing()
 
 add_library(port INTERFACE)
diff --git a/upb/json_decode.c b/upb/json_decode.c
index cc8139e..404ec1c 100644
--- a/upb/json_decode.c
+++ b/upb/json_decode.c
@@ -350,7 +350,7 @@ static char jsondec_escape(jsondec *d) {
 
 static uint32_t jsondec_codepoint(jsondec *d) {
   uint32_t cp = 0;
-  const char *end;
+  const char *end = NULL;
 
   if (d->end - d->ptr < 4) {
     jsondec_err(d, "EOF inside string");
@@ -926,8 +926,8 @@ static upb_msgval jsondec_msg(jsondec *d, const upb_fielddef *f) {
 
 static void jsondec_field(jsondec *d, upb_msg *msg, const upb_msgdef *m) {
   upb_strview name;
-  const upb_fielddef *f;
-  const upb_fielddef *preserved;
+  const upb_fielddef *f = NULL;
+  const upb_fielddef *preserved = NULL;
 
   name = jsondec_string(d);
   jsondec_entrysep(d);
@@ -1205,8 +1205,8 @@ static void jsondec_struct(jsondec *d, upb_msg *msg, const upb_msgdef *m) {
 static void jsondec_wellknownvalue(jsondec *d, upb_msg *msg,
                                    const upb_msgdef *m) {
   upb_msgval val;
-  const upb_fielddef *f;
-  upb_msg *submsg;
+  const upb_fielddef *f = NULL;
+  upb_msg *submsg = NULL;
 
   switch (jsondec_peek(d)) {
     case JD_NUMBER:
@@ -1262,7 +1262,7 @@ static upb_strview jsondec_mask(jsondec *d, const char *buf, const char *end) {
    * transform in place. */
   const char *ptr = buf;
   upb_strview ret;
-  char *out;
+  char *out = NULL;
 
   ret.size = end - ptr;
   while (ptr < end) {
@@ -1331,7 +1331,7 @@ static void jsondec_anyfield(jsondec *d, upb_msg *msg, const upb_msgdef *m) {
 static const upb_msgdef *jsondec_typeurl(jsondec *d, upb_msg *msg,
                                          const upb_msgdef *m) {
   const upb_fielddef *type_url_f = upb_msgdef_itof(m, 1);
-  const upb_msgdef *type_m;
+  const upb_msgdef *type_m = NULL;
   upb_strview type_url = jsondec_string(d);
   const char *end = type_url.data + type_url.size;
   const char *ptr = end;
@@ -1361,7 +1361,7 @@ static void jsondec_any(jsondec *d, upb_msg *msg, const upb_msgdef *m) {
   /* string type_url = 1;
    * bytes value = 2; */
   const upb_fielddef *value_f = upb_msgdef_itof(m, 2);
-  upb_msg *any_msg;
+  upb_msg *any_msg = NULL;
   const upb_msgdef *any_m = NULL;
   const char *pre_type_data = NULL;
   const char *pre_type_end = NULL;
