diff --git a/llvm/cmake/modules/LLVMConfig.cmake.in b/llvm/cmake/modules/LLVMConfig.cmake.in
index 2d90512bfb06..97aacd6f9c0f 100644
--- a/llvm/cmake/modules/LLVMConfig.cmake.in
+++ b/llvm/cmake/modules/LLVMConfig.cmake.in
@@ -51,16 +51,18 @@ set(LLVM_ENABLE_ASSERTIONS @LLVM_ENABLE_ASSERTIONS@)

 set(LLVM_ENABLE_EH @LLVM_ENABLE_EH@)

+include(CMakeFindDependencyMacro)
+
 set(LLVM_ENABLE_FFI @LLVM_ENABLE_FFI@)
 if(LLVM_ENABLE_FFI)
-  find_package(FFI)
+  find_dependency(FFI)
 endif()

 set(LLVM_ENABLE_RTTI @LLVM_ENABLE_RTTI@)

 set(LLVM_ENABLE_TERMINFO @LLVM_ENABLE_TERMINFO@)
 if(LLVM_ENABLE_TERMINFO)
-  find_package(Terminfo)
+  find_dependency(Terminfo)
 endif()

 set(LLVM_ENABLE_THREADS @LLVM_ENABLE_THREADS@)
@@ -70,25 +72,28 @@ set(LLVM_ENABLE_UNWIND_TABLES @LLVM_ENABLE_UNWIND_TABLES@)
 set(LLVM_ENABLE_ZLIB @LLVM_ENABLE_ZLIB@)
 if(LLVM_ENABLE_ZLIB)
   set(ZLIB_ROOT @ZLIB_ROOT@)
-  find_package(ZLIB)
+  find_dependency(ZLIB)
 endif()

 set(LLVM_ENABLE_ZSTD @LLVM_ENABLE_ZSTD@)
 if(LLVM_ENABLE_ZSTD)
-  find_package(zstd)
+  find_dependency(zstd)
 endif()

 set(LLVM_ENABLE_LIBXML2 @LLVM_ENABLE_LIBXML2@)
 if(LLVM_ENABLE_LIBXML2)
-  find_package(LibXml2)
+  find_dependency(LibXml2)
 endif()

 set(LLVM_ENABLE_CURL @LLVM_ENABLE_CURL@)
 if(LLVM_ENABLE_CURL)
-  find_package(CURL)
+  find_dependency(CURL)
 endif()

 set(LLVM_WITH_Z3 @LLVM_WITH_Z3@)
+if(LLVM_WITH_Z3)
+  find_dependency(Z3 4.7.1)
+endif()

 set(LLVM_ENABLE_DIA_SDK @LLVM_ENABLE_DIA_SDK@)

