From 6fa9c9acbce6bb5d0e8464951ab383e9776b6ca8 Mon Sep 17 00:00:00 2001
From: Peter Goodman <peter.goodman@gmail.com>
Date: Wed, 19 Apr 2023 13:41:58 -0400
Subject: [PATCH] Fixes to clang's tablegen of attributes.

---
 clang/utils/TableGen/ClangAttrEmitter.cpp | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/clang/utils/TableGen/ClangAttrEmitter.cpp b/clang/utils/TableGen/ClangAttrEmitter.cpp
index de608c780..e092aec8d 100644
--- a/clang/utils/TableGen/ClangAttrEmitter.cpp
+++ b/clang/utils/TableGen/ClangAttrEmitter.cpp
@@ -1233,6 +1233,18 @@ namespace {
     }
 
     void writeHasChildren(raw_ostream &OS) const override { OS << "true"; }
+
+    // PASTA PATCH: Check the optionality of an attribute.
+    std::string getIsOmitted() const override {
+      if (!isOptional()) {
+        return this->SimpleArgument::getIsOmitted();
+      }
+
+      std::string output;
+      llvm::raw_string_ostream OS(output);
+      OS << "(get" << getUpperName() << "() == nullptr)";
+      return output;
+    }
   };
 
   class VariadicExprArgument : public VariadicArgument {
-- 
2.40.0

